src/freenet/node/PeerLocation.java

```diff
@@ -107,4 +111,101 @@ synchronized double setLocation(double newLoc) {
 		return oldLoc;
 	}
 
+	/**
+	 * Finds the position of the first element in the sorted list greater than the given element,
+	 * or -1 if none.
+	 * This is a binary search of logarithmic complexity.
+	 */
+	private static int findFirstGreater(final double[] elems, final double x) {
+		int low = 0;
+		int high = elems.length;
+		int mid;
+		if (elems[high - 1] <= x) {
+			return -1;
+		}
+		while (low != high) {
+			mid = low + (high - low) / 2;
+			if (elems[mid] > x) {
+				high = mid;
+			} else {
+				low = mid + 1;
+			}
+		}
+		return low;
+	}
+
+	/**
+	 * Finds the position of the closest location in the sorted list of locations.
+	 * This is a binary search of logarithmic complexity.
+	 */
+	static int findClosestLocation(final double[] locs, final double l) {
+		assert(locs.length > 0);
+		if (locs.length == 1) {
+			return 0;
+		}
+		final int firstGreater = findFirstGreater(locs, l);
+		final int left;
+		final int right;
+		if (firstGreater == -1 || firstGreater == 0) {
+			// All locations are greater, or all locations are smaller.
+			// Closest location must be either smallest or greatest location.
+			left = locs.length - 1;
+			right = 0;
+		} else {
+			left = firstGreater - 1;
+			right = firstGreater;
+		}
+		if (Location.distance(l, locs[left]) <= Location.distance(l, locs[right])) {
+			return left;
```

Comment:
We're always approaching it from the *left* in the circular key space: we always take the solution that is, say, counterclockwise from the target (or is it actually clockwise? doesn't really matter). Whether or not we are wrapping around from 1 to 0 does not matter here. I named the variables `left` and `right` instead of `low` and `high` to avoid this exact confusion ☺

As such, this only matters when we route to a location exactly inbetween two FOAFs of a single peer, which happens practically never. The previous approach to resolving this choice was "whichever FOAF appeared earlier", the new approach is "whichever FOAF is counterclockwise to the target".